### PR TITLE
MAID-2418: upgrade web api

### PR DESCRIPTION
**Pull request by hunterlester**

### CODE REVIEW

File: src/api/app.js [hunk 1]
```diff
@@ -18,8 +18,7 @@ module.exports.manifest = {
   getOwnContainer: 'promise',
   getContainer: 'promise',
   reconnect: 'promise',
-  logPath: 'promise',
-  free: 'sync'
```
Comment (bochaco) [inline src/api/app.js]: We need to keep all `free` functions except for MutableData.free which doesn't exist anymore.

File: src/api/app.js [hunk 2]
```diff
@@ -44,8 +43,8 @@ module.exports.manifest = {
  *    .then((appHandle) => {
  *       console.log('SAFEApp instance initialised and handle returned: ', appHandle);
  *    });
- **/
```
Comment (bochaco) [inline src/api/app.js]: Please make sure this doesn't affect/break the documentation generator

File: src/api/app.js [hunk 3]
```diff
@@ -44,8 +43,8 @@ module.exports.manifest = {
...
-module.exports._with_async_cb_initialise = (appInfo, enableLog, safeAppGroupId) => {
+ */
+module.exports._with_async_cb_initialise = (appInfo, enableLog, safeAppGroupId) => { // eslint-disable-line no-underscore-dangle, max-len
```
Comment (bochaco) [inline src/api/app.js:53]: why are this eslint exceptions here now?

File: etc/documentation.yml [hunk 4]
```diff
@@ -101,42 +78,43 @@ toc:
...
     description: The safeCrypto API functions provide you with handy cryptographic functions
   - window.safeCrypto.sha3Hash
   - window.safeCrypto.getAppPubSignKey
   - window.safeCrypto.getAppPubEncKey
   - window.safeCrypto.generateEncKeyPair
-  - window.safeCrypto.getSignKeyFromRaw
-  - window.safeCrypto.pubEncKeyKeyFromRaw
-  - window.safeCrypto.secEncKeyKeyFromRaw
+  - window.safeCrypto.generateSignKeyPair
+  - window.safeCrypto.pubSignKeyFromRaw
```
Comment (bochaco) [inline etc/documentation.yml:111]: You are missing `window.safeCrypto.secSignKeyFromRaw`

File: src/api/crypto.js [hunk 5]
```diff
@@ -180,6 +214,23 @@ module.exports.generateEncKeyPairFromRaw = (appHandle, rawPublicKey, rawSecretKe
...
 
+/**
+ * Generate  a new signing key pair from raw secret and public keys
+ * @name window.safeCrypto.generateSignKeyPairFromRaw
+ *
+ * @param {SAFEAppHandle} appHandle
+ * @param {(String|Buffer)} rawPublicSignKey
+ * @param {(String|Buffer)} rawSecretSignKey
+ *
+ * @returns {Promise<SignKeyPair>}
```
Comment (bochaco) [inline src/api/crypto.js]: This should actually be `@returns {Promise<SignKeyPairHandle>}`

File: src/api/crypto_encryption_key_pair.js [hunk 6]
```diff
@@ -1,23 +1,22 @@
...
-  free: 'sync'
+  decryptSealed: 'promise'
 };
 
 /**
  * Get the Public Encryption Key instance of this key pair
- * @name window.safeCryptoKeyPair.getPubEncKey
+ * @name window.safeCryptoEncKeyPair.getPubEncKey
  *
  * @param {KeyPairHandle} keyPairHandle the KeyPair handle
```
Comment (bochaco) [inline src/api/crypto_encryption_key_pair.js]: This should actually be `@param {EncKeyPairHandle} encKeyPairHandle the EncryptionKeyPair handle` otherwise the doc won't link it correctly to the EncKeyPairHandle documentation.

File: src/api/crypto_encryption_key_pair.js [hunk 7]
```diff
@@ -28,15 +27,15 @@ module.exports.getPubEncKey = (keyPairHandle) => {
 
 /**
  * Get the Secrect Encryption Key instance of this key pair
- * @name window.safeCryptoKeyPair.getSecEncKey
+ * @name window.safeCryptoEncKeyPair.getSecEncKey
  *
  * @param {KeyPairHandle} keyPairHandle the KeyPair handle
```
Comment (bochaco) [inline src/api/crypto_encryption_key_pair.js]: This should actually be `@param {EncKeyPairHandle} encKeyPairHandle the EncryptionKeyPair handle` otherwise the doc won't link it correctly to the EncKeyPairHandle documentation.

File: src/api/crypto_encryption_key_pair.js [hunk 8]
```diff
@@ -47,7 +46,7 @@ module.exports.getSecEncKey = (keyPairHandle) => {
 
 /**
  * Decrypt the given ciphertext with a seal (buffer or string) using the private and public key
- * @name window.safeCryptoKeyPair.decryptSealed
+ * @name window.safeCryptoEncKeyPair.decryptSealed
  *
  * @param {KeyPairHandle} keyPairHandle the KeyPair handle
```
Comment (bochaco) [inline src/api/crypto_encryption_key_pair.js]: This should actually be `@param {EncKeyPairHandle} encKeyPairHandle the EncryptionKeyPair handle`, otherwise the doc won't link it correctly to the EncKeyPairHandle documentation.

File: src/api/crypto_public_sign_key.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
+const { getObj } = require('./helpers');
+
+module.exports.manifest = {
+  getRaw: 'promise',
+  verify: 'promise'
```
Comment (bochaco) [inline src/api/crypto_public_sign_key.js]: You are missing the `free` function.

File: src/api/crypto_secret_sign_key.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
+const { getObj } = require('./helpers');
+
+module.exports.manifest = {
+  getRaw: 'promise',
+  sign: 'promise'
+};
```
Comment (bochaco) [inline src/api/crypto_secret_sign_key.js]: As a reminder :) we need the `free` function here

File: src/api/mutable_data.js [hunk 11]
```diff
@@ -337,7 +317,7 @@ module.exports.decrypt = (mdHandle, value) => {
  *     console.log('Name: ', res.name.buffer);
  *     console.log('Tag: ', res.tag);
```
Comment (bochaco) [inline src/api/mutable_data.js]: The `tag` attribute was renamed to `type_tag`

File: src/api/mutable_data.js [hunk 12]
```diff
@@ -394,26 +374,7 @@ module.exports.get = (mdHandle, key) => {
  *
  * @returns {Promise} it resolves when finished creating it
  *
- * @example // Committing a MutableData to the network:
```
Comment (bochaco) [inline src/api/mutable_data.js]: Why are we removing this example snippet?
I f we adapt the code for the new permissions format the rest should all work fine

File: src/api/mutable_data.js [hunk 13]
```diff
@@ -447,17 +408,12 @@ module.exports.getEntries = (mdHandle) => {
  *
  * @param {MutableDataHandle} mdHandle the MutableData handle
  *
- * @returns {Promise<KeysHandle>} the Keys handle
+ * @returns {Promise<Array>} entry keys
  *
- * @example // Retrieving the keys:
```
Comment (bochaco) [inline src/api/mutable_data.js]: We should keep the example, and now we can also add a line which prints out the list since it's just a JS array now.

File: src/api/mutable_data.js [hunk 14]
```diff
@@ -466,17 +422,12 @@ module.exports.getKeys = (mdHandle) => {
  *
  * @param {MutableDataHandle} mdHandle the MutableData handle
  *
- * @returns {Promise<ValuesHandle>} the Values handle
+ * @returns {Promise<Array>} entry values
  *
- * @example // Retrieving the values:
```
Comment (bochaco) [inline src/api/mutable_data.js]: We should keep the example, and now we can also add a line which prints out the list since it's just a JS array now.

File: src/api/mutable_data.js [hunk 15]
```diff
@@ -557,30 +508,18 @@ module.exports.delUserPermissions = (mdHandle, signKeyHandle, version) => {
  *
  * @param {MutableDataHandle} mdHandle the MutableData handle
  * @param {SignKeyHandle|null} signKeyHandle the sign key to lookup for
- * @param {PermissionsSetHandle} pmSetHandle the PermissionsSet to set to
+ * @param {Array} pmSet - array of permissions
```
Comment (bochaco) [inline src/api/mutable_data.js]: Just for consistency, please remove the `-`

File: src/api/mutable_data.js [hunk 16]
```diff
@@ -557,30 +508,18 @@ module.exports.delUserPermissions = (mdHandle, signKeyHandle, version) => {
...
  * @param {Number} version the version successor, to confirm you are
  *        actually asking for the right state
  *
  * @returns {Promise} resolves when finished
  *
- * @example // Setting a new permission into a MutableData:
```
Comment (bochaco) [inline src/api/mutable_data.js]: We should keep it and just adapt it to the new permissions set format

File: src/api/mutable_data_entries.js [hunk 17]
```diff
@@ -66,8 +65,8 @@ module.exports.get = (entriesHandle, keyName) => {
  *          console.log('Version: ', v.version);
  *       }).then(_ => console.log('Iteration finished'))
  *    );
- **/
-module.exports._with_cb_forEach = (entriesHandle) => {
+*/
+module.exports._with_cb_forEach = (entriesHandle) => { // eslint-disable-line max-len, no-underscore-dangle
```
Comment (bochaco) [inline src/api/mutable_data_entries.js:70]: Why do we need the eslint exception here since it was ok before?

File: src/api/mutable_data_permissions.js [hunk 18]
```diff
@@ -54,63 +52,42 @@ module.exports.getPermissionsSet = (permissionsHandle, signKeyHandle) => {
  *
  * @param {PermissionsHandle} permissionsHandle the Permissions handle
  * @param {SignKeyHandle|null} signKeyHandle the sign key to map to
- * @param {PermissionsSetHandle} pmSetHandle - the permissions set you'd like insert
+ * @param {Array} permSetArray - array of permissions
  *
  * @returns {Promise} resolves once finished
  *
- * @example // Inserting a new permissions set into a MutableData:
```
Comment (bochaco) [inline src/api/mutable_data_permissions.js]: We keep the example and adapt it to the new permissions set format.

File: src/api/mutable_data_permissions.js [hunk 19]
```diff
@@ -54,63 +52,42 @@ module.exports.getPermissionsSet = (permissionsHandle, signKeyHandle) => {
...
- *    .then((permsHandle) => window.safeMutableDataPermissions.forEach(permsHandle, (p) => {
- *          console.log('Permissions entry handle: ', p);
- *       }).then(_ => console.log('Iteration finished'))
- *    );
- **/
-module.exports._with_cb_forEach = (permissionsHandle) => {
-  return forEachHelper(permissionsHandle, true);
+ *    .then((permsHandle) => window.safeMutableDataPermissions.listPermissionSets(permsHandle))
+ *    .then((permsArray) => {
+ *      console.log(permsArray);
```
Comment (bochaco) [inline src/api/mutable_data_permissions.js]: Please put some text ion the `console.log` just to make the example nicer/cleaner, e.g. console.log("Permission Sets retrieved:", permsArray);

**Review by bochaco — CHANGES_REQUESTED** [review]

File: src/api/app.js [hunk 20]
```diff
@@ -1,6 +1,6 @@
 const safeApp = require('@maidsafe/safe-node-app');
 const ipc = require('./ipc');
-const { genHandle, getObj, freeObj, netStateCallbackHelper } = require('./helpers');
+const { genHandle, getObj, netStateCallbackHelper } = require('./helpers');
 
 module.exports.manifest = {
```
Comment (bochaco) [inline src/api/app.js]: We are missing some new functions that are now exposed from safe_app_nodejs: `clearObjectCache`, `isMockBuild`, `getAccountInfo`, `isNetStateInit`, `isNetStateConnected` and `isNetStateDisconnected`.
We'd also have to exposed the constants, but that's a bit more complex and we can do it in a separate PR.

**Review by bochaco — CHANGES_REQUESTED** [review]

File: src/api/mutable_data_entries.js (hunk 17 above)
Comment (hunterlester) [reply src/api/mutable_data_entries.js:70]: When running `yarn lint` or `npm run lint` there are many lint errors, so I'm not sure what was happening before.

File: src/api/app.js [hunk 21]
```diff
@@ -293,6 +298,71 @@ module.exports.networkState = (appHandle) => {
...
+ * Resets the object cache kept by the underlyging library
+ * @name window.safeApp.clearObjectCache
+ *
+ * @param {SAFEAppHandle} appHandle
+ *
+ * @returns {Boolean}
+ * */
+module.exports.clearObjectCache = (appHandle) => {
+  return getObj(appHandle)
+  .then((obj) => obj.app.clearObjectCache());
```
Comment (bochaco) [inline src/api/app.js]: I think for this we also need to remove all the handles we have for the app, although we need to keep the app's handle.

File: src/api/crypto.js [hunk 22]
```diff
@@ -128,14 +130,19 @@ module.exports.pubSignKeyFromRaw = (appHandle, raw) => {
...
  * @param {(String|Buffer)} raw the raw input string
  *
- * @returns {Promise<SecSignKeyHandle>} 
+ * @returns {Promise<SecSignKeyHandle>}
  *
+ * @example // Interpret the secret sign key from raw buffer
+ * const signKeyPairHandle = await window.safeCrypto.generateSignKeyPair(appHandle);
+ * let secretSignKeyHandle = await window.safeaCryptoSignKeyPair.getSecSignKey(signKeyPairHandle);
+ * const rawSecretSignKey = await window.safeCryptoSecSignKey.getRaw(secretSignKeyHandle);
+ * secretSignKeyHandle = await window.safeCrypto.secSignKeyFromRaw(appHandle, rawSecretKey);
```
Comment (bochaco) [inline src/api/crypto.js]: I think this should be `secretSignKeyHandle = await window.safeCrypto.secSignKeyFromRaw(appHandle, rawSecretSignKey);`.
We'd need to translate all the example to use `await` to be consistent, we can do this separately.

File: src/api/app.js [hunk 23]
```diff
@@ -10,15 +10,20 @@ module.exports.manifest = {
   authoriseContainer: 'promise',
   authoriseShareMd: 'promise',
   webFetch: 'promise',
-  isRegistered: 'promise',
   networkState: 'promise',
+  isNetStateInit: 'sync',
+  isNetStateConnected: 'sync',
+  isNetStateDisconnected: 'sync',
+  clearObjectCache: 'promise',
+  isMockBuild: 'promise',
```
Comment (bochaco) [inline src/api/app.js]: These new functions need to also be listed in the `documentation.yml` file

**Review by bochaco — CHANGES_REQUESTED** [review]

File: src/api/app.js (hunk 23 above)
Comment (hunterlester) [reply src/api/app.js]: oooooh, that's why I can't get them to show up. thank you :-D

File: src/api/app.js (hunk 21 above)
Comment (hunterlester) [reply src/api/app.js]: All other requests have been address but I don't understand this one.

File: src/api/app.js [hunk 24]
```diff
@@ -306,6 +306,10 @@ module.exports.networkState = (appHandle) => {
  * @param {SAFEAppHandle} appHandle
  *
  * @returns {Boolean}
+ *
+ * @example // Checking if app's network state is INIT
+ * window.safeApp.isNetStateInit(appHandle)
+ *   .then((s) => console.log('App is in network state: ', s));
```
Comment (bochaco) [inline src/api/app.js]: This will print out a boolean rather than a state string

File: src/api/app.js [hunk 25]
```diff
@@ -345,6 +357,15 @@ module.exports.isNetStateDisconnected = (appHandle) => {
...
  *
  * @returns {Boolean}
+ *
+ * @example // Resets native object cache as well as front-end object handles except for SAFEApp
+ * window.safeApp.clearObjectCache(appHandle)
+ *   .then(() => {
+ *     console.log('All network objects dropped');
+ *     console.log(appHandle, ' still valid.);
+ *     window.safeApp.isRegistered(appHandle)
+ *     .then(console.log);
```
Comment (bochaco) [inline src/api/app.js]: Nice example, although this last line will log just a boolean, perhaps improve it to show a message that says something about that value.

**Review by bochaco — APPROVED** [review]
